### PR TITLE
[LV][AArch64] LoopVectorizer allows scalable frem instructions

### PR DESCRIPTION
**Pull request by paschalis-mpeis**
In AArch64, when an 'frem' instruction uses scalable vectors, it will be replaced with a vector library call. LoopVectorize is now aware of that so it no longer returns invalid costs.

When it is not scalable, it returns the default costs, which are delegated to the BaseT TTI Implementation.

### CODE REVIEW

File: llvm/lib/Analysis/TargetTransformInfo.cpp [hunk 1]
```diff
@@ -17,7 +17,6 @@
 #include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/Module.h"
 #include "llvm/IR/Operator.h"
-#include "llvm/IR/PatternMatch.h"
```
Comment (mgabka) [inline llvm/lib/Analysis/TargetTransformInfo.cpp:20]: looks like accidental change

File: llvm/test/Analysis/CostModel/AArch64/arith-fp-sve.ll [hunk 2]
```diff
@@ -137,14 +138,14 @@ define void @fdiv() {
...
 ; CHECK-LABEL: 'frem'
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V4F16 = frem <vscale x 4 x half> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V8F16 = frem <vscale x 8 x half> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V16F16 = frem <vscale x 16 x half> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V2F32 = frem <vscale x 2 x float> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V4F32 = frem <vscale x 4 x float> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V8F32 = frem <vscale x 8 x float> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V2F64 = frem <vscale x 2 x double> undef, undef
-; CHECK-NEXT:  Cost Model: Invalid cost for instruction: %V4F64 = frem <vscale x 4 x double> undef, undef
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 10 for instruction: %V4F16 = frem <vscale x 4 x half> undef, undef
```
Comment (mgabka) [inline llvm/test/Analysis/CostModel/AArch64/arith-fp-sve.ll:141]: in these tests veclib is not available so the cost should still be invalid

**Review by mgabka — CHANGES_REQUESTED** [review]

File: llvm/lib/Analysis/TargetTransformInfo.cpp (hunk 1 above)
Comment (paschalis-mpeis) [reply llvm/lib/Analysis/TargetTransformInfo.cpp:20]: This header is not used directly, so I thought of removing it.
It's unrelated to my patch though.

**Comment by paschalis-mpeis** [comment]
Forced pushed a041258 to additionally rebase to main.

CostModel is now computed in LoopVectorizer.

File: llvm/test/Analysis/CostModel/AArch64/arith-fp-sve.ll [hunk 3]
```diff
@@ -1,7 +1,8 @@
-; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
-; RUN: opt < %s -enable-no-nans-fp-math -passes="print<cost-model>" 2>&1 -disable-output -mtriple=aarch64 -mattr=+fullfp16 -mattr=+sve | FileCheck %s
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -mattr=+sve -mattr=+fullfp16 -enable-no-nans-fp-math -disable-output -passes="print<cost-model>" %s 2>&1 | FileCheck %s
```
Comment (fhahn) [inline llvm/test/Analysis/CostModel/AArch64/arith-fp-sve.ll:2]: unrelated changed?

File: llvm/test/Analysis/CostModel/AArch64/arith-fp.ll [hunk 4]
```diff
@@ -1,7 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
 ; RUN: opt < %s -enable-no-nans-fp-math  -passes="print<cost-model>" 2>&1 -disable-output -mtriple=aarch64 -mattr=+fullfp16 | FileCheck %s
 
-target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
+target triple = "aarch64-unknown-linux-gnu"
```
Comment (fhahn) [inline llvm/test/Analysis/CostModel/AArch64/arith-fp.ll:4]: unrelated changes?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 5]
```diff
@@ -7142,6 +7142,19 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
...
 
+    // Some targets replace frem with vector library calls.
+    if (I->getOpcode() == Instruction::FRem && VectorTy->isScalableTy()) {
+      LibFunc Func;
+      if (TLI->getLibFunc(I->getOpcode(), I->getType(), Func)) {
+        if (TLI->isFunctionVectorizable(TLI->getName(Func))) {
+          SmallVector<Type *, 4> OpTypes;
+          for (auto &Op : I->operands())
+            OpTypes.push_back(Op->getType());
+          return TTI.getCallInstrCost(nullptr, VectorTy, OpTypes, CostKind);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7153]: IIUC this doesn't account for the cost of the `frem` implementation itself. What about the case where a library provides `frem`, but it is more expensive?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 6]
```diff
@@ -7142,6 +7142,19 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
         Legal->isInvariant(Op2))
       Op2Info.Kind = TargetTransformInfo::OK_UniformValue;
 
+    // Some targets replace frem with vector library calls.
+    if (I->getOpcode() == Instruction::FRem && VectorTy->isScalableTy()) {
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7146]: Why limit to scalable types?

**Comment by fhahn** [comment]
We are already checking if a library function exists, wouldn't it be more accurate to replace to widen recipe with a WidenCall recipe with the function?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 5 above)
Comment (paschalis-mpeis) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Now it compares against the default cost. Since this is now in place, the restriction to scalable types (in the [next comment](https://github.com/llvm/llvm-project/pull/76247#discussion_r1441661371) is lifted.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 7]
```diff
@@ -6951,10 +6951,26 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
...
+    auto InstrCost = TTI.getArithmeticInstrCost(
         I->getOpcode(), VectorTy, CostKind,
         {TargetTransformInfo::OK_AnyValue, TargetTransformInfo::OP_None},
         Op2Info, Operands, I);
+
+    // Some targets replace frem with vector library calls.
+    if (I->getOpcode() == Instruction::FRem) {
+      LibFunc Func;
+      if (TLI->getLibFunc(I->getOpcode(), I->getType(), Func)) {
+        if (TLI->isFunctionVectorizable(TLI->getName(Func))) {
```
Comment (mgabka) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:6963]: I think you could just combine this condition with the one above to avoid excessive indentation.

File: llvm/test/Analysis/CostModel/AArch64/arith-fp-sve-frem.ll [hunk 8]
```diff
@@ -0,0 +1,98 @@
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py UTC_ARGS: --filter "cost.*vscale.*frem" --version 4
+
+; RUN: opt -mattr=+sve -passes=loop-vectorize -debug-only=loop-vectorize -disable-output -S < %s 2>&1 | FileCheck %s --check-prefix=NO-VECLIB
```
Comment (mgabka) [inline llvm/test/Analysis/CostModel/AArch64/arith-fp-sve-frem.ll:3]: could you also addd RUN lines for Neon, i.e without +sve? That wilol require changing the file name as well.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 9]
```diff
@@ -6951,10 +6951,26 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
...
+          SmallVector<Type *, 4> OpTypes;
+          for (auto &Op : I->operands())
+            OpTypes.push_back(Op->getType());
+          auto CallCost =
+              TTI.getCallInstrCost(nullptr, VectorTy, OpTypes, CostKind);
+          return std::min(InstrCost, CallCost);
```
[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:6969]: @mgabka If CallCost is invalid because there's no available mapping or very high because it's suboptimal and the target in question (not AArch64, at least at present) has an frem instruction, we would want to return InstrCost.

If replace-with-veclib were to override that, I think it would be a bug in that pass.

**Review by mgabka — APPROVED** [review]
It LGTM, but probably worth to wait some time to give a chance @huntergr-arm and @fhahn to also have a look.

**Review by huntergr-arm — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (paschalis-mpeis) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: The below PR will bring such functionality to `replace-with-veclib`. Thank you all for your input!
- #78688
